node/src/Graph.cpp

```diff
@@ -165,53 +67,43 @@ namespace node {
-            Napi::Object jsResource = jsResources.Get(name).As<Napi::Object>();
-            // dictionary Input {
-            //   required ArrayBufferView buffer;
-            //   sequence<long> dimensions;
-            // };
-            // dictionary Output {
-            //   ArrayBufferView buffer;
-            //   sequence<long> dimensions;
-            // };
             T resource = {};
-            if (!jsResource.Has("data")) {
-                // Input buffer is required.
-                return false;
-            }
-            int jsElementLength = 0;
-            if (jsResource.Has("data")) {
-                if (!jsResource.Get("data").IsTypedArray()) {
+            std::string name = names.Get(i).As<Napi::String>().Utf8Value();
+            // FIXME: validate the type of typed array.
+            Napi::TypedArray jsTypedArray;
+            if (jsResources.Get(name).IsTypedArray()) {
+                jsTypedArray = jsResources.Get(name).As<Napi::TypedArray>();
+            } else {
+                Napi::Object jsResource = jsResources.Get(name).As<Napi::Object>();
+                if (!jsResource.Has("resource") || !jsResource.Has("dimensions")) {
+                    // Input buffer and dimensions are required.
                     return false;
                 }
+                if (!jsResource.Get("resource").IsTypedArray()) {
+                    return false;
+                }
+                jsTypedArray = jsResource.Get("resource").As<Napi::TypedArray>();
 
-                // FIXME: validate the type of typed array.
-
-                Napi::TypedArray jsTypedArray = jsResource.Get("data").As<Napi::TypedArray>();
-                resource.buffer = reinterpret_cast<void*>(
-                    reinterpret_cast<int8_t*>(jsTypedArray.ArrayBuffer().Data()) +
-                    jsTypedArray.ByteOffset());
-                resource.size = jsTypedArray.ByteLength();
-                jsElementLength = jsTypedArray.ElementSize();
-            }
-            if (HasOptionMember(jsResource, "dimensions")) {
                 if (!GetInt32Array(jsResource.Get("dimensions"), resource.dimensions)) {
                     return false;
                 }
-
-                // Check dimensions.
-                if (jsElementLength) {
-                    int dimensionSize = 1;
-                    for (auto& dim : resource.dimensions) {
-                        dimensionSize *= dim;
-                    }
-                    if (dimensionSize != jsElementLength) {
-                        return false;
-                    }
+                if (SizeOfShape(resource.dimensions) != jsTypedArray.ElementSize()) {
+                    return false;
                 }
             }
+            resource.buffer = reinterpret_cast<void*>(
+                reinterpret_cast<int8_t*>(jsTypedArray.ArrayBuffer().Data()) +
+                jsTypedArray.ByteOffset());
+            resource.byteLength = jsTypedArray.ByteLength();
```

Comment:
Set `byteOffset` as we support it now
```c++
resource.byteOffset = jsTypedArray.ByteOffset();
```

Reply:
The `resource.buffer ` is used as base line?

```
resource.buffer = reinterpret_cast<void*>(
                reinterpret_cast<int8_t*>(jsTypedArray.ArrayBuffer().Data()) +
                jsTypedArray.ByteOffset());
```

=>

`resource.buffer = reinterpret_cast<void*>(reinterpret_cast<int8_t*>(jsTypedArray.ArrayBuffer().Data()));`

Reply:
Constant also need to support MLBufferView, how about supporting it with another PR?

"name": "constant",
        "returns": "operand",
        "args": [
          {"name": "desc", "type": "operand descriptor", "annotation": "const*"},
          {"name": "value", "type": "void", "annotation": "const*"},
          {"name": "size", "type": "size_t"}
        ]


=>

 
"name": "constant",
        "returns": "operand",
        "args": [
          {"name": "desc", "type": "operand descriptor", "annotation": "const*"},
          {"name": "value", "type": "array buffer view"}
        ]